feat: schedule support cron

zxbyoyoyo opened the pull request:
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. Close issue syntax: `Fixes #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

Fixes 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

zxbyoyoyo committed:
feat: schedule support cron

bowenliang123 commented on api/extensions/ext_celery.py:
```diff
@@ -50,21 +50,48 @@ def __call__(self, *args: object, **kwargs: object) -> object:
+    """Determine the schedule type based on configuration."""
+
+    # Fetch configuration values
+    scheduler_type = app_config.get("CELERY_BEAT_SCHEDULER_TYPE")  # Options: 'time' or 'cron'
```
Use `dify_config.CELERY_BEAT_SCHEDULER_TYPE` directly for accessing the config value.

bowenliang123 commented on api/extensions/ext_celery.py:
```diff
@@ -50,21 +50,48 @@ def __call__(self, *args: object, **kwargs: object) -> object:
+
+    # Fetch configuration values
+    scheduler_type = app_config.get("CELERY_BEAT_SCHEDULER_TYPE")  # Options: 'time' or 'cron'
+    day = app_config.get("CELERY_BEAT_SCHEDULER_TIME")
+    cron_expression = app_config.get("CELERY_BEAT_SCHEDULER_CRON_EXPRESSION")
```
ditto.

bowenliang123 commented on api/extensions/ext_celery.py:
```diff
@@ -50,21 +50,48 @@ def __call__(self, *args: object, **kwargs: object) -> object:
+    cron_expression = app_config.get("CELERY_BEAT_SCHEDULER_CRON_EXPRESSION")
+
+    # Determine the schedule based on the 'scheduler_type' value
+    if scheduler_type == "cron" and cron_expression:
```
How about use `match scheduler_type : case` for guarding condiction of scheduler_type. 

And check the cron_expression inside the `cron` branch, throwing error when cron_expression is missing. It's way better than just falling to unsatisfied branch of other type.

zxbyoyoyo committed:
fix: enhance error messages in get_schedule

- Added checks for missing configurations.
- Improved cron expression format validation.

bowenliang123 commented on api/extensions/ext_celery.py:
```diff
@@ -37,34 +37,74 @@ def __call__(self, *args: object, **kwargs: object) -> object:
+                raise ValueError(
+                    "Configuration Error: 'CELERY_BEAT_SCHEDULER_CRON_EXPRESSION' is required when using the 'cron' scheduler type.")
+            try:
+                minute, hour, day_of_month, month_of_year, day_of_week = cron_expression.split()
```
How about use [cron-validator](https://pypi.org/project/cron-validator/) to check and parse the units rather than just spliting the string in spaces?

bowenliang123 reviewed (approved):
LGTM overall. cc @laipz8200 @takatost

laipz8200 commented:
Please sync the code with the main branch to make the checks pass.
